### PR TITLE
Fixed #29727 -- Made nonexistent joins in F() raise FieldError.

### PR DESCRIPTION
**Pull request by alexh546**
Fetching the value of a non-existing joined field on a model using ``F()`` does not raise an error. Instead, it is just giving the primary key of the fetched object.

This patch ensures a ``FieldError`` when last element in a name passed to ``F()`` is neither a field nor a transform.

[#29727](https://code.djangoproject.com/ticket/29727)

### CODE REVIEW

**Comment by pope1ni** [comment]
This requires a Trac ticket to describe the problem you are trying to fix.
Please see the [contributing guidelines](https://docs.djangoproject.com/en/dev/internals/contributing/).

**Comment by alexh546** [comment]
Sorry, forgot to put a reference to the ticket in the PR.

File: tests/expressions/tests.py [hunk 1]
```diff
@@ -591,6 +591,14 @@ def test_pickle_expression(self):
         expr.convert_value  # populate cached property
         self.assertEqual(pickle.loads(pickle.dumps(expr)), expr)
 
+    def test_incorrect_field_expression(self):
+        with self.assertRaisesMessage(FieldError, "Cannot resolve keyword 'nope' into field."):
+            list(Employee.objects.filter(firstname=F('nope')))
```
Comment (felixxm) [inline tests/expressions/tests.py]: I agree that `test_incorrect_field_expression()` should be in the `BasicExpressionsTests` instead of `ExpressionsNumericTests`, but this change is not related with a fix. Please move this to a separate commit.

File: tests/expressions/tests.py [hunk 2]
```diff
@@ -591,6 +591,14 @@ def test_pickle_expression(self):
...
+
+    def test_incorrect_field_expression_in_join(self):
+        with self.assertRaisesMessage(FieldError, "Cannot resolve keyword 'nope' into field."):
+            list(Company.objects.filter(ceo__firstname=F('point_of_contact__nope')))
```
Comment (felixxm) [inline tests/expressions/tests.py]: Running `test_incorrect_field_expression_in_join()` without a fix, doesn't behave like it's described in the ticket on the PostgreSQL database. It raises: 

_"django.db.utils.ProgrammingError: operator does not exist: character varying = integer"_

IMO using `ceo__pk` instead of `ceo_firstname` should fix this issue.

File: django/db/models/sql/query.py [hunk 3]
```diff
@@ -1594,10 +1594,12 @@ def resolve_ref(self, name, allow_joins=True, reuse=None, summarize=False, simpl
         else:
             field_list = name.split(LOOKUP_SEP)
             join_info = self.setup_joins(field_list, self.get_meta(), self.get_initial_alias(), can_reuse=reuse)
-            targets, _, join_list = self.trim_joins(join_info.targets, join_info.joins, join_info.path)
+            targets, final_alias, join_list = self.trim_joins(join_info.targets, join_info.joins, join_info.path)
             if len(targets) > 1:
                 raise FieldError("Referencing multicolumn fields with F() objects "
                                  "isn't supported")
+            # Trigger FieldError when last field doesn't resolve to a field or a transform
```
Comment (felixxm) [inline django/db/models/sql/query.py]: Maybe: _"Ensure the last element is a field or a transform."_ :thinking:

File: tests/expressions/tests.py (hunk 1 above)
Comment (alexh546) [reply tests/expressions/tests.py]: Agreed.

PS. I take this opportunity to see what happens to comments when rewriting history in PR, haven't done that before.

**Comment by alexh546** [comment]
@felixxm thanks for the review, I broke it when I rewrote history but I've made the changes you suggested. Not sure why the first commit displays failed checks though :-/

**Comment by alexh546** [comment]
@felixxm Ok I think I'm done

**Comment by felixxm** [comment]
@alexh546 Thanks :+1: . IMO it is ready for the final review. Can you add a reference to the ticket in the last commit message? _"Fixed #29727 -- ..."_
